File: python/mirage/commands/builtin/redis/wc/wc.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

from collections.abc import AsyncIterator

from mirage.accessor.redis import RedisAccessor
from mirage.cache.index import IndexCacheStore
from mirage.commands.builtin.aggregators import wc_aggregate
from mirage.commands.builtin.redis._provision import file_read_provision
from mirage.commands.builtin.utils.stream import _resolve_source
from mirage.commands.registry import command
from mirage.commands.spec import SPECS
from mirage.core.redis.glob import resolve_glob
from mirage.core.redis.read import read_bytes
from mirage.io.types import ByteSource, IOResult
from mirage.types import PathSpec


async def _wc_lines_stream(
        source: AsyncIterator[bytes]) -> AsyncIterator[bytes]:
    count = 0
    async for chunk in source:
        count += chunk.count(b"\n")
    yield str(count).encode()


@command("wc",
         resource="redis",
         spec=SPECS["wc"],
         aggregate=wc_aggregate,
         provision=file_read_provision)
async def wc(
    accessor: RedisAccessor,
    paths: list[PathSpec],
    *texts: str,
    stdin: AsyncIterator[bytes] | bytes | None = None,
    args_l: bool = False,
    w: bool = False,
    c: bool = False,
    m: bool = False,
    L: bool = False,
    index: IndexCacheStore = None,
    **_extra: object,
) -> tuple[ByteSource | None, IOResult]:
    if paths and accessor.store is not None:
        paths = await resolve_glob(accessor, paths, index)
        outputs: list[str] = []
        total_lines = total_words = total_bytes = 0
        for p in paths:
            data = await read_bytes(accessor, p)
            text = data.decode(errors="replace")
            line_count = text.count("\n")
            word_count = len(text.split())
            byte_count = len(data)
            if L:
                max_len = max((len(ln) for ln in text.splitlines()), default=0)
                outputs.append(f"{max_len}\t{p.original}")
            elif args_l:
                outputs.append(f"{line_count}\t{p.original}")
                total_lines += line_count
            elif w:
                outputs.append(f"{word_count}\t{p.original}")
                total_words += word_count
            elif c:
                outputs.append(f"{byte_count}\t{p.original}")
                total_bytes += byte_count
            elif m:
                char_count = len(text)
                outputs.append(f"{char_count}\t{p.original}")
                total_bytes += char_count
            else:
                outputs.append(
                    f"{line_count}\t{word_count}\t{byte_count}\t{p.original}")
                total_lines += line_count
                total_words += word_count
                total_bytes += byte_count
        if len(paths) > 1:
            if args_l:
                outputs.append(f"{total_lines}\ttotal")
            elif w:
                outputs.append(f"{total_words}\ttotal")
            elif c:
                outputs.append(f"{total_bytes}\ttotal")
            elif m:
                outputs.append(f"{total_bytes}\ttotal")
            else:
                outputs.append(
                    f"{total_lines}\t{total_words}\t{total_bytes}\ttotal")
        return "\n".join(outputs).encode(), IOResult()

    source: AsyncIterator[bytes] = _resolve_source(stdin,
                                                   "wc: missing operand")

    if args_l:
        return _wc_lines_stream(source), IOResult()

    raw = b""
    async for chunk in source:
        raw += chunk
    text = raw.decode(errors="replace")
    lc = text.count("\n")
    wc_val = len(text.split())
    bc = len(raw)
    cc = len(text)

    if L:
        max_len = max((len(ln) for ln in text.splitlines()), default=0)
        return str(max_len).encode(), IOResult()
    if w:
        return str(wc_val).encode(), IOResult()
    if m:
        return str(cc).encode(), IOResult()
    if c:
        return str(bc).encode(), IOResult()
    return f"{lc}\t{wc_val}\t{bc}".encode(), IOResult()

File: typescript/packages/core/src/commands/builtin/s3/wc.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import type { S3Accessor } from '../../../accessor/s3.ts'
import { resolveGlob } from '../../../core/s3/glob.ts'
import { read as s3Read } from '../../../core/s3/read.ts'
import { IOResult, materialize, type ByteSource } from '../../../io/types.ts'
import { ResourceName, type PathSpec } from '../../../types.ts'
import { command, type CommandFnResult, type CommandOpts } from '../../config.ts'
import { specOf } from '../../spec/builtins.ts'
import { resolveSource } from '../utils/stream.ts'
import { fileReadProvision } from './provision.ts'

const ENC = new TextEncoder()
const DEC = new TextDecoder('utf-8', { fatal: false })

async function* wcLinesStream(source: AsyncIterable<Uint8Array>): AsyncIterable<Uint8Array> {
  let count = 0
  for await (const chunk of source) {
    for (let i = 0; i < chunk.byteLength; i++) if (chunk[i] === 0x0a) count += 1
  }
  yield ENC.encode(String(count))
}

function countChar(text: string, ch: string): number {
  let n = 0
  for (const c of text) if (c === ch) n += 1
  return n
}

function maxLineLength(text: string): number {
  let max = 0
  let current = 0
  for (const c of text) {
    if (c === '\n') {
      if (current > max) max = current
      current = 0
    } else {
      current += 1
    }
  }
  if (current > max) max = current
  return max
}

async function wcCommand(
  accessor: S3Accessor,
  paths: PathSpec[],
  _texts: string[],
  opts: CommandOpts,
): Promise<CommandFnResult> {
  const f = opts.flags
  const lFlag = f.args_l === true
  const wFlag = f.w === true
  const cFlag = f.c === true
  const mFlag = f.m === true
  const LFlag = f.L === true
  if (paths.length > 0) {
    const resolved = await resolveGlob(accessor, paths, opts.index ?? undefined)
    const first = resolved[0]
    if (first === undefined) return [null, new IOResult()]
    const data = await s3Read(accessor, first, opts.index ?? undefined)
    const text = DEC.decode(data)
    const lineCount = countChar(text, '\n')
    const wordCount = text.split(/\s+/).filter((s) => s !== '').length
    const byteCount = data.byteLength
    if (LFlag) return [ENC.encode(String(maxLineLength(text))), new IOResult()]
    if (lFlag) return [ENC.encode(String(lineCount)), new IOResult()]
    if (wFlag) return [ENC.encode(String(wordCount)), new IOResult()]
    if (mFlag) return [ENC.encode(String(text.length)), new IOResult()]
    if (cFlag) return [ENC.encode(String(byteCount)), new IOResult()]
    const out = `${String(lineCount)}\t${String(wordCount)}\t${String(byteCount)}`
    return [ENC.encode(out), new IOResult()]
  }
  let source: AsyncIterable<Uint8Array>
  try {
    source = resolveSource(opts.stdin, 'wc: missing operand')
  } catch (err) {
    const msg = err instanceof Error ? err.message : String(err)
    return [null, new IOResult({ exitCode: 1, stderr: ENC.encode(`${msg}\n`) })]
  }
  if (lFlag) {
    const out: ByteSource = wcLinesStream(source)
    return [out, new IOResult()]
  }
  const raw = await materialize(source)
  const text = DEC.decode(raw)
  const lc = countChar(text, '\n')
  const wcVal = text.split(/\s+/).filter((s) => s !== '').length
  const bc = raw.byteLength
  const cc = text.length
  if (LFlag) return [ENC.encode(String(maxLineLength(text))), new IOResult()]
  if (wFlag) return [ENC.encode(String(wcVal)), new IOResult()]
  if (mFlag) return [ENC.encode(String(cc)), new IOResult()]
  if (cFlag) return [ENC.encode(String(bc)), new IOResult()]
  return [ENC.encode(`${String(lc)}\t${String(wcVal)}\t${String(bc)}`), new IOResult()]
}

export const S3_WC = command({
  name: 'wc',
  resource: ResourceName.S3,
  spec: specOf('wc'),
  fn: wcCommand,
  provision: fileReadProvision,
})
